### PR TITLE
atom parsing, now using the apache abdera library 

### PR DESCRIPTION
**Pull request by janvanmansum**
Fixes EASY-1130
#### When applied it will
- Replace the code that used XPath to retrieve information from the deposit receipt and statement with code that uses [Apache Abdera](https://cwiki.apache.org/confluence/display/ABDERA/Creating+and+Consuming+Atom+Documents) 
#### Where should the reviewer @DANS-KNAW/easy start?

Some place where Abdera is used, for example:
https://github.com/DANS-KNAW/easy-sword2-dans-examples/compare/master...janvanmansum:EASY-1130?expand=1#diff-9f2c5221b830f5da12154e6b07b6afd5R70
#### How should this be manually tested?

Run the examples.

### CODE REVIEW

**Comment by janvanmansum** [comment]
@DANS-KNAW/easy please review.

File: src/main/java/nl/knaw/dans/easy/sword2examples/Common.java [hunk 1]
```diff
@@ -59,44 +63,49 @@ public static String readEntityAsString(HttpEntity entity) throws IOException {
...
-     *        the XPath expression
-     * @return the result as a string
-     * @throws Exception
-     */
-    public static String getStringFromXml(String xml, String expr) throws Exception {
-        InputSource xmlSource = new InputSource(new StringReader(xml));
-        XPathFactory xpathFactory = XPathFactory.newInstance();
-        XPath xpath = xpathFactory.newXPath();
-        return (String) xpath.evaluate(expr, xmlSource, XPathConstants.STRING);
+    public static Entry parseEntry(String text) {
```
Comment (rvanheest) [inline src/main/java/nl/knaw/dans/easy/sword2examples/Common.java]: This method and `parseFeed` look exactly the same, except for their return type. These can be one method with a generic parameter.

File: src/main/java/nl/knaw/dans/easy/sword2examples/Common.java [hunk 2]
```diff
@@ -59,44 +63,49 @@ public static String readEntityAsString(HttpEntity entity) throws IOException {
...
+        Document<Feed> receipt = parser.parse(new StringReader(text));
+        return receipt.getRoot();
+    }
+
+    static void trackDeposit(CloseableHttpClient http, IRI statIri) throws Exception {
         CloseableHttpResponse response;
         String bodyText;
         System.out.println("Start polling Stat-IRI for the current status of the deposit, waiting 10 seconds before every request ...");
-        String state = null;
+        String stateTerm = null;
```
Comment (rvanheest) [inline src/main/java/nl/knaw/dans/easy/sword2examples/Common.java]: `null` assignment is not needed according to IntelliJ in this and the next line:

> the value is always overwritten with another assignment before the next variable read

File: src/main/java/nl/knaw/dans/easy/sword2examples/Common.java [hunk 3]
```diff
@@ -59,44 +63,49 @@ public static String readEntityAsString(HttpEntity entity) throws IOException {
...
+        String stateText = null;
         while (true) {
             Thread.sleep(10000);
             System.out.print("Checking deposit status ... ");
-            response = http.execute(new HttpGet(statIri));
+            response = http.execute(new HttpGet(statIri.toURI()));
```
Comment (rvanheest) [inline src/main/java/nl/knaw/dans/easy/sword2examples/Common.java]: Changing the type of `statIri` into a `URI` in the input parameter saves a lot of computations that are actually equal for each run through the while loop. As you get an `IRI` from Abdera, you can do this conversion `toURI` in different places.

File: src/main/java/nl/knaw/dans/easy/sword2examples/Common.java [hunk 4]
```diff
@@ -59,44 +63,49 @@ public static String readEntityAsString(HttpEntity entity) throws IOException {
...
             bodyText = readEntityAsString(response.getEntity());
-            state = getStringFromXml(bodyText,
-                    "//*[local-name() = 'category' and @scheme = 'http://purl.org/net/sword/terms/state' and @label = 'State']/@term");
-            System.out.println(state);
-            if (state.equals("INVALID") || state.equals("REJECTED") || state.equals("FAILED")) {
+            Feed statement = parseFeed(bodyText);
+            List<Category> states = statement.getCategories("http://purl.org/net/sword/terms/state");
+            if (states.size() < 1) {
```
Comment (rvanheest) [inline src/main/java/nl/knaw/dans/easy/sword2examples/Common.java]: `states.size() < 1` can better be `states.isEmpty`. Refer to https://pavelfatin.com/scala-collections-tips-and-tricks/#sequences-length

File: src/main/java/nl/knaw/dans/easy/sword2examples/Common.java (hunk 2 above)
Comment (rvanheest) [reply src/main/java/nl/knaw/dans/easy/sword2examples/Common.java]: (10 minutes later...) Actually, now that I look at it again, you don't need these declarations here at all!

File: src/main/java/nl/knaw/dans/easy/sword2examples/Common.java [hunk 5]
```diff
@@ -59,44 +63,49 @@ public static String readEntityAsString(HttpEntity entity) throws IOException {
...
+                System.err.println("ERROR: NO STATE FOUND");
+                System.exit(1);
+            } else if (states.size() > 1) {
+                System.err.println("ERROR: FOUND TOO MANY STATES (" + states.size() + "). CAN ONLY HANDLE ONE");
+                System.exit(1);
+            }
+            stateTerm = states.get(0).getTerm();
+            stateText = states.get(0).getText();
```
Comment (rvanheest) [inline src/main/java/nl/knaw/dans/easy/sword2examples/Common.java]: only compute this in the `else-if` branch

File: src/main/java/nl/knaw/dans/easy/sword2examples/Common.java [hunk 6]
```diff
@@ -118,11 +127,11 @@ public static CloseableHttpClient createHttpClient(URI iri, String uid, String p
         return HttpClients.custom().setDefaultCredentialsProvider(credsProv).build();
     }
 
-    public static CloseableHttpResponse sendChunk(DigestInputStream dis, int size, String method, URI iri, String filename, String mimeType, CloseableHttpClient http, boolean inProgress) throws Exception {
+    public static CloseableHttpResponse sendChunk(DigestInputStream dis, int size, String method, IRI iri, String filename, String mimeType, CloseableHttpClient http, boolean inProgress) throws Exception {
```
Comment (rvanheest) [inline src/main/java/nl/knaw/dans/easy/sword2examples/Common.java]: This `IRI` is also converted to a `URI`, as well as all its uses in the other classes. Why use this `IRI` anyway?

**Review by rvanheest — COMMENTED** [review]
A couple of comments...

After I wrote them, I realized that it would be just as easy to put them in [a PR](https://github.com/janvanmansum/easy-sword2-dans-examples/pull/1).

**Pull request merged**